src/Factory/RedirectFactory.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Setono\SyliusRedirectPlugin\Factory;
+
+use Setono\SyliusRedirectPlugin\Model\RedirectInterface;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Symfony\Component\Routing\RouterInterface;
+
+final class RedirectFactory implements RedirectFactoryInterface
+{
+    /** @var FactoryInterface */
+    private $decoratedFactory;
+
+    /** @var RouterInterface */
+    private $router;
+
+    public function __construct(FactoryInterface $decoratedFactory, RouterInterface $router)
+    {
+        $this->decoratedFactory = $decoratedFactory;
+        $this->router = $router;
+    }
+
+    public function createNew(): RedirectInterface
+    {
+        /** @var RedirectInterface $redirect */
+        $redirect = $this->decoratedFactory->createNew();
+
+        return $redirect;
+    }
+
+    public function createNewWithValues(
+        string $source,
+        string $destination,
+        bool $permanent = true,
+        bool $only404 = true
+    ): RedirectInterface {
+        $redirect = $this->createNew();
+
+        $redirect->setSource($source);
+        $redirect->setDestination($destination);
+        $redirect->setOnly404($only404);
+        $redirect->setPermanent($permanent);
+        $redirect->enable();
+
+        return $redirect;
+    }
+
+    public function createNewForProduct(
+        ProductInterface $product,
+        string $source,
+        string $destination,
+        bool $permanent = true,
+        bool $only404 = true
+    ): RedirectInterface {
+        $source = $this->router->generate('sylius_shop_product_show', ['slug' => $source]);
```

Comment:
I have thought a lot about injecting the router into the factory. Something tells me that the callee should input the correct source and destination without the factory needs to have a router.

Also this method doesn't use nothing but the channels from the product, which could have been inputted in a more generic way, i.e.:

```php
public function createNewWithValues(
        string $source,
        string $destination,
        bool $permanent = true,
        bool $only404 = true,
        array $channels = []
    ): RedirectInterface
```

Wdyt?

Reply:
Indeed. It's a small change that removes the factory dependency to the router. I'll change calling services and simply remove the `createNewForProduct` method.